### ISSUE TITLE
I hope IBus can keep my input when I switch between different engines.

### ISSUE THREAD

**Issue by qiangzhao** [ibus/ibus#1949]
Steps:
1, I input some Roma characters.
2, I found I'm working with an incorrect engine, So I switch to the one I really want.
3, Check with current input result:

Actually Result:
all input has been cleared.

Expect Result:
There will have a switch on IBus preference setup program about if you want to keep input content, if I select it, all my input characters will automatically redirect to the new engine.

Thanks!

**Comment by fujiwarat** [comment]
I don't understand totally what you mentioned.

> 2, I found I'm working with an incorrect engine, So I switch to the one I really want.

Which engine did you work with and which one did you switch to?

Which desktop do you use? GNOME?
And which versions?

**Comment by Zhaor** [comment]
>>> Which engine did you work with and which one did you switch to?
There's no specific engine, or may I say with any different engine.
For example, 
1). I work with "ibus-anthy", 
2). I input 4 characters "Cool",  these characters are in a pre-edit box now but *NOT* transport to editor program(just like "gedit").
3). Suddenly, I find what I really want is to input English, but not Japanese. 
4). So I use compose key to switch input to English.
5). I hope these characters have been automatically input. the current situation is just be cleared.

Which desktop do you use? GNOME?
Yes, GNOME.

And which versions?
3.26

But I suppose IBUS can keep and transport these pre-edit characters,  it's better to not to depend on desktop env.

Thanks!

**Comment by fujiwarat** [comment]
Thank you for your explanation.
I understood you mean to commit the preedit text when you switch engines.
Normally I use the switching shortcut key to clear the mistaken preedit text so I didn't expect such a usage.

Currently I recommend to use a shortcut key to switch input modes (E.g. Hiragana and direct input) rather than a shortcut key to switch engines (E.g. anthy and xkb:us::eng) since we switch the two modes frequently and switching engines is slower than switching modes.
So I think asking the feature is to implement in an IBus engine instead of IBus core.

E.g. ibus-anthy has a setting of "Behavior on Focus Out: Commit" in /usr/lib*/ibus-setup-anthy and the engine can commit the preedit text when the focus is changed and I think the similar feature can be implemented.

I mean if you wish the feature in ibus-anthy, probably I can implement it in ibus-anthy but if you wish it in others, please file the issue in those bug trackers.

**Comment by fujiwarat** [comment]
> And I suppose IBUS can keep and transport these pre-edit characters, it's better to not to depend on desktop env. Thanks!

I agree with you. I asked it to reproduce your problem and didn't understand your problem.

**Comment by qiangzhao** [comment]
Indeed, switch between modes is faster and smoother than switch engines.
But we should aware that many engines only support limited modes:
eg, switch Japanese to Traditional Chinese 
      switch Japanese to German.
In this scenario, the user only can switch engine to accomplish the goal.

Another point is, I think this work could be done inside the input frame.
the pre-edit text is controlled by ibus-daemon, and switch action is also finished by IBUS,
So IBUS has the ability to serve all engines to accomplish this. 
and modify IBUS is better than modifying so many engines 1 by 1.

Regards!

**Comment by fujiwarat** [comment]
Right. I also think many CJK engines switches the two modes and , I can understand some people might like to commit the preedit text in that scenario since the two modes are switched frequently.
However I think many people don't care to commit the preedit in that scenario and I even think when engines are switched, the engine's preedit and lookup table is better to be reset and the current state is fine at the moment.
I think ibus-daemon wont commit preedit text because each engine can commit the text and it would be a conflict feature.
I wonder why you don't commit the preedit text before you switch engines.

**Comment by qiangzhao** [comment]
>>> Right. I also think many CJK engines switches the two modes and , I can understand some people might like to commit the preedit text in that scenario since the two modes are switched frequently.
However I think many people don't care to commit the preedit in that scenario and I even think when engines are switched, the engine's preedit and lookup table is better to be reset and the current state is fine at the moment.

Yes, this is a user's personal taste problem, different people will have a different choice.
That's just why I suggest to add an option at ibus-perference-setup, "keep pre-edit text" or not. With that selection, people can get what they want.

>>> I think ibus-daemon won't commit pre-edit text because each engine can commit the text and it would be a conflicting feature.

At least, I think ibus will provide an interface, say, 
int ibus_bus_send_preedit_to_buffer (
        IBusBus *bus,
        gchar *pedit_char;
        GCancellable *cancellable); 

int ibus_bus_get_preedit_buffer (
        IBusBus *bus,
        gchar *pedit_string;
        GCancellable *cancellable); 
Maybe you have a more better name and proper parameters, It's just an example. With these interfaces, engines can have a unique mechinasm to do this work.

>>> I wonder why you don't commit the preedit text before you switch engines.

Because I suppose to commit before switch engines are very hard. there are too many engines, every engine needs to modify if we choose this way. But if we modify ibus-daemon, there's only 1 place to work with.
like: "anthy", "mozc", "kkc", "skk", "cangjie", "rime", "chew", "m17n", "hangl", "pingyin", "libpinyin", "zhuyin", they all have pre-edit, and CJK people may create more engine in future, never knows.

Regards!

**Comment by fujiwarat** [comment]
Do you use ibus-anthy or a Japanese engine daily? Since Japanese preedit text could not be completed without lookup table in many cases, committing text would be always a natural process and I'm thinking Japanese users would avoid such a feature, i.e. reset the preedit text and lookup table.
I'm thinking it's good to triage your languages only and every case won't be fixed at the moment until other users request the same feature.
It is still not clarified for me which ibus engines yourself are using currently and why you wish .to commit the preedit automatically when you switch engines.

> 2). I input 4 characters "Cool", these characters are in a pre-edit box now but NOT transport to editor program(just like "gedit").

In that case, ibus-anthy would shows "cおおl" in Hiragana mode and I think it's a wrong preedit and should be reset.

> 2, I found I'm working with an incorrect engine, So I switch to the one I really want.

Actually you refers you used an incorrect engine and if you switch an incorrect engine to a right engine, I think the preedit should be reset.

**Comment by fujiwarat** [comment]
> In that case, ibus-anthy would shows "cおおl" in Hiragana mode and I think it's a wrong preedit and should be reset.

I see you typed Shift key but the word "Cool" only would be correct accidentally and the sentence could not be completed in the current engine. In that case, myself would even type Escape key and switch to the correct engine. And I set the default input mode is "Latin" in ibus-setup-anthy and the problem would not happen.

**Comment by qiangzhao** [comment]
> Do you use ibus-anthy or a Japanese engine daily? Since Japanese preedit text could not be completed without lookup table in many cases, committing text would be always a natural process and I'm thinking Japanese users would avoid such a feature, i.e. reset the preedit text and lookup table.

if so, This feature will only be needed for Chinese engines.

> I'm thinking it's good to triage your languages only and every case won't be fixed at the moment until other users request the same feature.

Okay, Fine.

> It is still not clarified for me which ibus engines yourself are using currently and why you wish .to commit the preedit automatically when you switch engines.

I use pinyin and wubi mostly, but I also test other input engines. :)

**Comment by fujiwarat** [comment]
> And I set the default input mode is "Latin" in ibus-setup-anthy and the problem would not happen.

As I noted, I think yourself issue can be fixed with the setting of the default input mode of ibus-anthy at the moment.
As you use pinyin, wubi mostly and other IMEs sometimes and I think your typos mostly happen with Japanese IMEs.

**Comment by oxr463** [comment]
Going back to the original description, is it possible to use Ctrl + A to select the text so that it can be copied before changing the engine? Currently, if I enter some text using anthy, I cannot use Ctrl + A to select it, (and obviously Ctrl + C to copy), before switching back to english.

**Comment by fujiwarat** [comment]
@oxr463 If you use ibus-anthy, you can use `Ctrl-a` when you switch the "Input Mode" to Latin mode but you cannot use `Ctrl-a` with Hiragana mode. YOu can switch the Input Mode on the ibus panel menu or a shortcut key.

**Comment by oxr463** [comment]
Hmm.. Why doesn't it work with Hiragana mode?

**Comment by fujiwarat** [comment]
@oxr463 I confirmed your mention is valid for MS-IME too. Please file your issue to https://github.com/ibus/ibus-anthy/issues
